### ISSUE TITLE
FBX import now working

### ISSUE THREAD

**Issue by jayrulez** [godotengine/godot#46906]
**Godot version:**
3.2.4 RC3

**OS/device including version:**
Windows 10

**Issue description:**
The imported model is horribly broken.

**Steps to reproduce:**
Import the attached model in a godot project.

This is what the model looks like in Godot:
![image](https://user-images.githubusercontent.com/78807/110839735-e338d500-8271-11eb-9168-e5accecd5467.png)

For reference, this is the same model imported in stride:
![image](https://user-images.githubusercontent.com/78807/110840522-cf41a300-8272-11eb-9fab-a096986be5b9.png)

Flax:
![image](https://user-images.githubusercontent.com/78807/110840923-4d9e4500-8273-11eb-83cf-0d1c4413e502.png)

What it should look like:
![image](https://user-images.githubusercontent.com/78807/110841040-6e669a80-8273-11eb-884e-24450636fa80.png)

Here's the model for debug purposes:
[50005011.zip](https://github.com/godotengine/godot/files/6125415/50005011.zip)

**Comment by Calinou** [comment]
Which software was used to export the model? Cinema4D FBX isn't supported yet.

**Comment by jayrulez** [comment]
I don't know what DCC they are from. I'll ask the creator.

I get the same result as you see there on 3.2.3.

 Godot uses Assimp right? Is the Assimp version in Godot updated?
Another engine that uses Assimp is able to import it properly so I think Godot is doing something wrong somehow.
![image](https://user-images.githubusercontent.com/78807/110852876-e5566000-8280-11eb-944d-f2be5da08d48.png)

Have you considered https://github.com/nem0/openfbx for fbx imports in Godot?
It seems to provide very good results importing fbx in my experience.

**Comment by Calinou** [comment]
> Godot uses Assimp right? Is the Assimp version in Godot updated?

Godot never used Assimp directly (in any released version at least). It uses a reverse engineered implementation of FBX with some code taken from Assimp (but modified to be usable in Godot).

**Comment by jayrulez** [comment]
> > Godot uses Assimp right? Is the Assimp version in Godot updated?
> > Another engine that uses Assimp is able to import it properly so I think Godot is doing something wrong somehow.
> 
> Godot never used Assimp directly (in any stable version at least). It uses a reverse engineered implementation of FBX with some code taken from Assimp (but modified to be usable in Godot).
> 
> > Have you considered nem0/openfbx for fbx imports in Godot?
> > It seems to provide very good results importing fbx in my experience.
> 
> We have a renewed FBX importer in 3.2.4: https://godotengine.org/article/fbx-importer-rewritten-for-godot-3-2-4
I see. I hope it will improve with future versions then as the last one I tried 3.2.4 RC4 is still broken.

I don't think I can do anymore that providing the sample model in this case.

If these ones can't be supported then let me know when you get an answer on that.

Thanks

**Comment by RevoluPowered** [comment]
Looks like it was made in 3DS Max, I'm fairly certain that the breakage is caused by geometric pivots in the FBX file.

The algorithm for max was never implemented, we did Maya's format only. 

I'll look into this when I can, as I have actually had them working fine in an older branch, the hard part though is making the animations after this work correctly compensating for the geometric pivot.
 
 Pivot formula we use currently is statically applying Maya's equation:
 https://help.autodesk.com/view/FBX/2017/ENU/?guid=__files_GUID_10CDD63C_79C1_4F2D_BB28_AD2BE65A02ED_htm
 
 In order to support this in Godot the following methods could be used: 

1.) apply the geometric pivot to the node then affine it on the child nodes
```cpp
foreach( node in scene )
{
    node->transform *= geometric pivot;
    foreach( children under node ) 
    {
        child_node->transform *= geometric_pivot.inverse();
    }
}
```
2.) Apply pivot transform to every vertex on the mesh

I'll check the geometric pivot is to blame in the meantime, but i had a spaceship model do the exact same thing because of the same issue.

**Comment by snougo** [comment]
the model directly casue a crash
3.2.4 rc4  MacOS 10.14.6
![Screen Shot 2021-03-12 at 22 10 42](https://user-images.githubusercontent.com/64188282/110951287-ddc9b080-837f-11eb-92a6-f497b7dc7709.png)

**Comment by rijx** [comment]
Has Godot considered using https://github.com/nem0/OpenFBX ?

I heard many good things about it. Flax engine uses it.

**Comment by Calinou** [comment]
@rijx See above: https://github.com/godotengine/godot/issues/46906#issuecomment-797044056

**Comment by mrParabol** [comment]
A model that I exported from blender as fbx is not imported properly in godot. For a proper import in godot, what format should I export in blender?

**Comment by Calinou** [comment]
> A model that I exported from blender as fbx is not imported properly in godot. For a proper import in godot, what format should I export in blender?

glTF 2.0 is recommended.

**Comment by wcbx** [comment]
> @rijx See above: [#46906 (comment)](https://github.com/godotengine/godot/issues/46906#issuecomment-797044056)

I think you should look at the evidence here. This person was able to import their FBX into everything else but Godot and it worked fine. So if you're not considering moving to something else for importing FBX you should be ready to admit that quite obviously something in the FBX importer is broken.

**Comment by AgentSmith0** [comment]
It think this issue is similar to this one: https://github.com/godotengine/godot/issues/45930

**Comment by shafnaz** [comment]
My Godot also crashes as soon as I drag & drop this .fbx into Godot.
It doesn't startup unless I remove the file from the project's folder through explorer.
This file contains 6 transform animated baked keyframes objects 54 frames long. It also contains blend shape animations.
Using Maya 2022 exported with FBX binary 2020.

[number_four.zip](https://github.com/godotengine/godot/files/8380094/number_four.zip)

**Comment by shafnaz** [comment]
Using Godot 3.4.4 Stable by the way

**Comment by akien-mga** [comment]
This is fixed by #59653.

https://user-images.githubusercontent.com/4701338/160845626-0498c395-7e5b-44b4-b44b-1fe168210475.mp4

For the `number_four.zip` MRP, it also seems fixed. The material import might not work 100% as intended, feel free to open a bug report after testing with the next 4.0 alpha (alpha 6 will have this change).

**Comment by Zireael07** [comment]
When's alpha 6 coming?

**Comment by akien-mga** [comment]
Two weeks after alpha 5, like each alpha blog post mentions :)

**Comment by jayrulez** [comment]
> This is fixed by #59653.
> 
>  simplescreenrecorder-2022-03-30_15.27.23.mp4 
> For the `number_four.zip` MRP, it also seems fixed. The material import might not work 100% as intended, feel free to open a bug report after testing with the next 4.0 alpha (alpha 6 will have this change).

Nice. Will this be backported to 3.x?

**Comment by Calinou** [comment]
> Nice. Will this be backported to 3.x?

Probably not, as this relies on the new import system which is backwards-incompatible.

**Comment by shafnaz** [comment]
Sounds great. Will we be still able to load our games properly in the new 4.0?

**Comment by Calinou** [comment]
> Sounds great. Will we be still able to load our games in the new 4.0 without changes?

No, as compatibility was broken in the `master` branch a long time ago.

**Comment by akien-mga** [comment]
I think this could be backported. Needs someone to try :)
